wallet/src/lib.rs

```diff
@@ -0,0 +1,383 @@
+        }
+    }
+
+    pub fn balance(&self, pk: PublicKey) -> Option<Value> {
+        let balance = self
+            .pk_index
+            .get(&pk)?
+            .iter()
+            .map(|id| self.utxos[id].note.value)
+            .sum();
+
+        Some(balance)
+    }
+
+    pub fn apply_block(&self, known_keys: &HashSet<PublicKey>, block: &WalletBlock) -> Self {
+        let mut utxos = self.utxos.clone();
+        let mut pk_index = self.pk_index.clone();
+
+        // Process each transaction in the block
+        for ledger_tx in &block.ledger_txs {
+            // Remove spent UTXOs (inputs)
+            for spent_id in &ledger_tx.inputs {
+                if let Some(utxo) = utxos.get(spent_id) {
+                    let pk = utxo.note.pk;
+                    utxos = utxos.remove(spent_id);
+
+                    if let Some(note_set) = pk_index.get(&pk) {
+                        let updated_set = note_set.remove(spent_id);
+                        if updated_set.is_empty() {
+                            pk_index = pk_index.remove(&pk);
+                        } else {
+                            pk_index = pk_index.insert(pk, updated_set);
+                        }
+                    }
+                }
+            }
+
+            // Add new UTXOs (outputs) - only if they belong to our known keys
+            for utxo in ledger_tx.utxos() {
+                if known_keys.contains(&utxo.note.pk) {
+                    let note_id = utxo.id();
+                    utxos = utxos.insert(note_id, utxo);
+
+                    let note_set = pk_index
+                        .get(&utxo.note.pk)
+                        .cloned()
+                        .unwrap_or_else(rpds::HashTrieSetSync::new_sync)
+                        .insert(note_id);
+                    pk_index = pk_index.insert(utxo.note.pk, note_set);
+                }
+            }
+        }
+
+        Self { utxos, pk_index }
+    }
+
+    /// Removes Utxos that do not contribute to meeting the `amount` threshold.
+    ///
+    /// As an example, suppose we hold notes valued [3 NMO, 4 NMO] and we asked
+    /// for 4 NMO then, since we sort the notes by value, we would have
+    /// first added the 3 NMO note and then then 4 NMO note to the selected
+    /// utxos list.
+    ///
+    /// The 4 NMO note alone would have satisfied the request, the 3 NMO note is
+    /// redundant and would be returned as change in a transaction.
+    ///
+    /// To resolve this, we remove as many of the smallest notes as we can while
+    /// still keep us above the requested amount.
+    fn remove_redundant_utxos(amount: Value, mut sorted_utxos: Vec<Utxo>) -> Vec<Utxo> {
+        debug_assert!(sorted_utxos.is_sorted_by_key(|utxo| utxo.note.value));
+
+        let mut skip_count = 0;
+        let mut temp_amount: Value = sorted_utxos.iter().map(|u| u.note.value).sum();
+
+        for utxo in &sorted_utxos {
```

Comment:
nit: you could also iterate in reverse and stop when it's more than the requested value but this works too

Reply:
true, I wrote both out and it's basically the same logic, I didn't see a complexity reduction there.

I'll keep this one as is